barretenberg/cpp/pil/avm/avm_alu.pil

```diff
@@ -474,3 +476,75 @@ namespace avm_alu(256);
     (p_sub_b_lo' - res_lo) * rng_chk_sel'= 0;
     (p_sub_b_hi' - res_hi) * rng_chk_sel'= 0;
 
+    // ========= SHIFT LEFT/RIGHT OPERATIONS ===============================
+    // Given inputs to a shift operation, a & b, and a memory tag, mem_tag.
+    // Split a into Big Endian hi and lo limbs, a_hi and a_lo, and the number of bits represented by the memory tag, t.
+    // QUESTION: SHOULD B BE CONSTRAINED TO BE U8 -> i.e. when would we shift more than 255 bits when the max number of bits of a is 128bits?
+    //
+    // === IF SHR
+    // (1) Prove the correct decomposition: a_hi * 2**b + a_lo = a
+    // (2) Range check a_hi < 2**(t-b) && a_lo < 2**b, ensure we have not overflowed the limbs during decomp
+    // (3) Return a_hi
+    //
+    //   <--(t-b) --> | Length: b
+    // ---------------|-----------------------
+    // |      a_hi    |          a_lo        | --> a
+    // ---------------------------------------
+    //
+    // === IF SHL
+    // (1) Prove the correct decomposition: a_hi * 2**(t-b) + a_lo = a
+    // (2) Range check a_hi < 2**b && a_lo < 2**(t-b)
+    // (3) Return a_lo * 2**b
+    // 
+    //          <-- b-->      |   <-- (t-b) -->
+    // -----------------------|---------------
+    // |      a_hi            |  a_lo        | --> a
+    // ---------------------------------------
+
+    // TODO: Variable Range Checks
+    // Check that a_lo and a_hi combine to be the same number of bits represented by the Memory Tag
+    // Can we can convert this into a fixed range check over b that might be more optimal for us? 
+    // SHR: a_hi < 2**(t - b) && a_lo < 2**b
+    // SHL: a_hi < 2**b && a_lo < 2**(t - b)
+
+    // === General Notes:
+    // There are probably ways to merge various relations for the SHL/SHR, but theyre separate
+    // now while we are still figuring out.
+
+    // We would like to calculate 2 ** ib, but it's costly in the circuit and if ib is <= U8
+    // we could just lookup the result of 2**b based on expanding the 8-bit lookup in main.
+    pol commit two_pow_b; 
+    pol commit pow_2_sel;
+    pow_2_sel = op_shl + op_shr;
+
+    
+    // Bit length, t, based on the mem tag.
+    pol MAX_BITS = u8_tag * 8 + 
+                u16_tag * 16 +
+                u32_tag * 32 +
+                u64_tag * 64 +
+                u128_tag * 128;
+
+    // The value of (t - b) bits
+    pol commit t_sub_b_bits;
+    t_sub_b_bits = MAX_BITS - ib;
```

Comment:
Needs to ensure somehow that ib < MAX_BITS. This might be guaranteed by the lookup which would fail?